### ISSUE TITLE
Counter-intuitive behavior of DisposeSessionResult

### ISSUE THREAD

**Issue by thibault-reigner** [neo4j/neo4j-dotnet-driver#326]
Currently, disposing a session automatically triggers the consumption (and in-memory storing !) of any remaining results : 

```csharp
        private void DisposeSessionResult()
        {
            if (_connection == null)
            {
                // there is no session result resources to dispose
                return;
            }

            if (_connection.IsOpen)
            {
                try
                {
                    // this will force buffering of all unconsumed result
                    _connection.Sync();
                }
                catch (Exception e)
                {
                    throw new ClientException(
                        $"Error when pulling unconsumed session.run records into memory in session: {e.Message}", e);
                }
                finally
                {
                    // there is a possibility that when error happens e.g. ProtocolError, the resources are not closed.
                    DisposeConnection();
                }
            }
            else
            {
                DisposeConnection();
            }
        }
```

If I am not mistaken it is because in the Bolt protocol, as soon as PULL_ALL has been started, there is no way to inform the server that the pulling can be cancelled.

However, for a use on production, where lazy evaluated / streaming based code is required it is not sustainable to perform this kind of operation since it is detrimental for performances (time and memory used to retrieved results which will be discarded anyway), if for some reason (exception, external event, ...) the enumeration is stopped.
For instance it can lead this type of code :

```
var query = session.Run("myQuery").Single(); 
```

to retrieve all the results and possibly throw an exception of type OutOfMemory rather than the actual 'more than one element in the collection'.

So far I have tested to "fix" this issue by editing the DisposeSessionResult function as follow : 

```
        private void DisposeSessionResult()
        {
            if (_connection == null)
            {
                // there is no session result resources to dispose
                return;
            }

            DisposeConnection();
        }
```

It seems to work perfectly but I have trouble to actually assess the impact of it.

Do you think this could be an optional behavior of the driver to perform this kind of operation on session's disposal instead of the current one ?

**Comment by zhenlineo** [comment]
Hi @thibault-reigner,

Have you tried with the method [`Statement#Consume`](https://github.com/neo4j/neo4j-dotnet-driver/blob/1.7/Neo4j.Driver/Neo4j.Driver/V1/IStatementResult.cs#L55)?

```
using(var session = driver.Session())
{
    result = session.Run(myQuery);
    try 
    {
        for(var record: result) 
        {
            // handle each record
        }
    }
    finally
    {
        result.consume(); 
        // the full result still streamed to client, 
        // but immediately discarded rather than buffered.
    }
}
```
Will this solve your problem?

**Comment by thibault-reigner** [comment]
@zhenlineo I will give it a try but it solves only a part of the problem : the memory consumption. At the end there is still network and processor time wasted by the consuming process.

**Comment by zhenlineo** [comment]
@thibault-reigner Right, we will think about this feature in future releases.

**Comment by zhenlineo** [comment]
We are looking into a react API for this driver which will probably solve the problem totally.

**Comment by technige** [comment]
Hi @thibault-reigner 

Bolt itself does not (yet) support flow control. This is on our roadmap, but for the time being, the only option available at protocol level is to stream back the entire result.

This therefore requires applications to be aware of this limitation and plan queries well enough ahead, wherever possible, such that discarding large parts of a result is an exceptional, not routine, activity. Cypher itself offers the `LIMIT` clause, which may be useful and the driver API allows you to `consume` the remainder of a result, as @zhenlineo mentioned. I'm not sure I see your "memory consumption" or "processor time" arguments here, since the remaining data is simply flushed from the incoming buffer, it's neither stored nor processed. Granted, there is more network activity than is ideal in this case.

I'd like to understand more about your particular use case and the impact on your application. Maybe we can help with query tuning or application design to help work around the absence of flow control. Do you also have some measurements that show excessive resource consumption? These could help us understand where your bottlenecks are.

**Comment by thibault-reigner** [comment]
@technige Thank you for the explanations.
 As I suspected in my first message, and you confirmed, the issue is in the Bolt protocol for the moment (no flow control).
Regarding the memory consumption, as of today, buffering everything and then discarding the whole buffer is the default behavior of the stream returned by Run (hence creating a potentially massive memory pick). So though it is possible to write extra-code as @zhenlineo, I don't really get the added value of not making this solution the default one.
Regarding the network time (and extremely minor processor time) I was referring to all the resources both from the server side and client side lost to maintain the stream of remaining results.

Note this is a very general statement, not an actual issue I have encountered in an application. Actually it would be much more like a reason to delay any 'go to production' process since all the "limit" kind of hacks are not reassuring, especially if I was planning to write a near real-time application processing a stream of results. Any exception requiring cancellation of the stream processing and needing to be time-bounded could not passively wait for the remaining results to all be consumed.

**Comment by technige** [comment]
What kind of Cypher queries are you running?

**Comment by thibault-reigner** [comment]
@technige Sorry for the delay. For the moment I am only using some basic queries like 
MATCH pattern WHERE stuff RETURN key nodes of the pattern
But on the long run I think it would be some more complex detection done through extensions, hence the queries would become : 
CALL procedure 
with eventually some YIELD + cypher post-process.

**Comment by technige** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Flow Control, Back Pressure & Reactive Drivers](https://trello.com/c/Q74jCr78/132-flow-control-back-pressure-reactive-drivers)

**Comment by technige** [comment]
Closing this issue as flow control has now been implemented in the reactive drivers (4.0).

https://www.nuget.org/packages/Neo4j.Driver/4.0.0
